src/clusterfuzz/_internal/bot/tasks/utasks/__init__.py

```diff
@@ -53,17 +55,70 @@ def _timestamp_now() -> Timestamp:
   return ts
 
 
-def _record_e2e_duration(start: Timestamp, utask_module, job_type: str,
-                         subtask: _Subtask, mode: _Mode):
-  duration_secs = (time.time_ns() - start.ToNanoseconds()) / 10**9
-  monitoring_metrics.UTASK_E2E_DURATION_SECS.add(
-      duration_secs, {
-          'task': task_utils.get_command_from_module(utask_module.__name__),
-          'job': job_type,
-          'subtask': subtask.value,
-          'mode': mode.value,
-          'platform': environment.platform(),
-      })
+class _MetricRecorder(contextlib.AbstractContextManager):
+  """Records task execution metrics, even in case of error and exceptions.
+
+  Members:
+    start_time_ns (int): The time at which this recorder was constructed, in
+      nanoseconds since the Unix epoch.
+  """
+
+  def __init__(self, subtask: _Subtask, mode: _Mode):
+    self.start_time_ns = time.time_ns()
+    self._subtask = subtask
+    self._mode = mode
+    self._labels = None
+
+    if subtask == _Subtask.PREPROCESS:
+      self._preprocess_start_time_ns = self.start_time_ns
+    else:
+      self._preprocess_start_time_ns = None
+
+  def set_task_details(self,
+                       utask_module,
+                       job_type: str,
+                       platform: str,
+                       preprocess_start_time: Optional[Timestamp] = None):
+    """Sets task details that might not be known at instantation time.
+
+    Must be called once for metrics to be recorded when exiting the context.
+
+    Args:
+      utask_module: The Python module corresponding to the task being executed.
+      job_type: The name of the job against which the task is being executed.
+      platform: The platform we are executing on, as given by
+        `environment.platform()`.
+      preprocess_start_time: Timestamp at which the preprocess subtask for
+        this task started executing, possibly in a different process. Must be
+        specified iff the subtask is not `Subtask.PREPROCESS`.
+    """
+    self._labels = {
+        'task': task_utils.get_command_from_module(utask_module.__name__),
+        'job': job_type,
+        'subtask': self._subtask.value,
+        'mode': self._mode.value,
+        'platform': platform,
+    }
+
+    if preprocess_start_time is not None:
```

Comment:
Why do we need to handle preprocess differently here?

Reply:
Because in preprocess, we don't have a `uworker_input` yet, much less its `preprocess_start_time` field, so we can't pass that value in here. In fact, `_MetricRecorder` is in charge of recording the start time as `self.start_time_ns`, which we then use to populate `uworker_input.preprocess_start_time`. We *could* pass that value back into `set_task_details()`, but it seems weird to allow potential errors.